lib/modules/manager/npm/artifacts.ts

```diff
@@ -0,0 +1,91 @@
+import { logger } from '../../../logger';
+import { exec } from '../../../util/exec';
+import type { ExecOptions } from '../../../util/exec/types';
+import { readLocalFile } from '../../../util/fs';
+import { regEx } from '../../../util/regex';
+import type { UpdateArtifact, UpdateArtifactsResult } from '../types';
+
+// eg. 8.15.5+sha256.4b4efa12490e5055d59b9b9fc9438b7d581a6b7af3b5675eb5c5f447cee1a589
+const versionWithHashRegString = '^(?<version>.*)\\+(?<hash>.*)';
+
+// Execute 'corepack use' command for npm manager updates
+// This step is necessary because Corepack recommends attaching a hash after the version
+// The hash is generated only after running 'corepack use' and cannot be fetched from the npm registry
+export async function updateArtifacts({
+  packageFileName,
+  config,
+  updatedDeps,
+  newPackageFileContent: existingPackageFileContent,
+}: UpdateArtifact): Promise<UpdateArtifactsResult[] | null> {
+  logger.debug(`npm.updateArtifacts(${packageFileName})`);
+  const packageManagerUpdate = updatedDeps.find(
+    (dep) => dep.depType === 'packageManager',
+  );
+
+  if (!packageManagerUpdate) {
+    logger.debug('No packageManager updates - returning null');
+    return null;
+  }
+
+  const { currentValue, depName, newVersion, rangeStrategy } =
+    packageManagerUpdate;
+
+  // Execute 'corepack use' command only if:
+  // 1. The package manager version is pinned in the project, or
+  // 2. The package manager version already has a hash appended
+  if (
+    !currentValue ||
+    (rangeStrategy !== 'pin' &&
+      !regEx(versionWithHashRegString).test(currentValue))
```

Comment:
check for current digest?

Reply:
We do not extract digest for package managers, so that filed is undefined.

Reply:
Should I instead, extract the hash as digest during extraction and use `pinDigests` to run `corepack` instead of `rangeStrategy`?

This didn't make sense to me in the beginning, but now it seems like the correct apporach.

Reply:
IMO we should not use `pinDigests` or `rangeStrategy:pin` to decide whether to append the hashes or not, because:

1. `rangeStrategy:pin`: just wrong since we are not pinning a range
2. `pinDigests`: the hash is not a regular digest, which we fetch from a regsitry

If a hash is present already, we append a hash. Else, we do not. This seems like the best appraoch to me ie. let the user decide for themselves, we just follow their choice.

Reply:
@rarkins @secustor WDYT?

Reply:
I think pinDigests makes sense for this capability. But I'm happy to document that this doesn't work with pinDigests in this PR and add a new low priority issue for its support if anyone wants to do it